specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2018-06-01-preview/apimdiagnostics.json

```diff
@@ -0,0 +1,533 @@
+      "properties": {
+        "samplingType": {
+          "type": "string",
+          "description": "Sampling type.",
+          "enum": [
+            "fixed"
+          ],
+          "x-ms-enum": {
+            "name": "SamplingType",
+            "modelAsString": true,
+            "values": [
+              {
+                "value": "fixed",
+                "description": "Fixed-rate sampling."
+              }
+            ]
+          }
+        },
+        "percentage": {
+          "type": "number",
+          "format": "double",
+          "minimum": 0,
+          "maximum": 100,
+          "description": "Rate of sampling for fixed-rate sampling."
+        }
+      },
+      "description": "Sampling settings for Diagnostic."
+    },
+    "PipelineDiagnosticSettings": {
+      "properties": {
+        "request": {
+          "allOf": [
+            {
+              "$ref": "#/definitions/HttpMessageDiagnostic"
+            }
+          ],
+          "description": "Diagnostic settings for request."
+        },
+        "response": {
+          "allOf": [
+            {
+              "$ref": "#/definitions/HttpMessageDiagnostic"
+            }
+          ],
+          "description": "Diagnostic settings for response."
+        }
+      },
+      "description": "Diagnostic settings for incoming/outcoming HTTP messages to the Gateway."
+    },
+    "HttpMessageDiagnostic": {
+      "properties": {
+        "headers": {
+          "type": "array",
+          "items": {
+            "type": "string"
+          },
+          "description": "Array of HTTP Headers to log."
+        },
+        "body": {
+          "allOf": [
+            {
+              "$ref": "#/definitions/BodyDiagnosticSettings"
+            }
+          ],
+          "description": "Body logging settings."
+        }
+      },
+      "description": "Http message diagnostic settings."
+    },
+    "BodyDiagnosticSettings": {
+      "properties": {
+        "bytes": {
+          "type": "integer",
+          "format": "int32",
+          "maximum": 8192,
+          "description": "Number of request body bytes to log."
+        }
+      },
+      "description": "Body logging settings."
+    },
+    "DiagnosticContractProperties": {
+      "properties": {
+        "alwaysLog": {
+          "type": "string",
+          "enum": [
+            "allErrors"
+          ],
+          "x-ms-enum": {
+            "name": "AlwaysLog",
+            "modelAsString": true,
+            "values": [
+              {
+                "value": "allErrors",
+                "description": "Always log all erroneous request regardless of sampling settings."
+              }
+            ]
+          },
+          "description": "Specifies for what type of messages sampling settings should not apply."
+        },
+        "loggerId": {
+          "type": "string",
+          "description": "Resource Id of a target logger."
+        },
+        "sampling": {
+          "allOf": [
```

Comment:
why "allOf" and why not a simple "$ref"

Reply:
same for others which define only one $ref

---
In reply to: [189126276](https://github.com/Azure/azure-rest-api-specs/pull/3066#discussion_r189126276) [](ancestors = 189126276)

Reply:
fixed